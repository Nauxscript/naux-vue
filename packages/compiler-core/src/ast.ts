export enum NodeTypes {
  TEXT,
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  ROOT,
}

export const TO_DISPLAY_STRING = 'toDisplayString'

export enum ElementTagTypes {
  START,
  END,
}
