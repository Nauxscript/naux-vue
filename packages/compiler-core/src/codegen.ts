export function generate(ast) {
  const context = createCodegenContext()
  const { push } = context
  const functionName = 'render'
  const signtures = ['_ctx', '_cache']
  push('return ')
  push(`function ${functionName}`)
  push(`(${signtures.join(',')}){`)
  getNode(ast, context)
  push('}')

  return {
    code: context.code,
  }
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source
    },
  }
  return context
}

function getNode(ast: any, context) {
  const { push } = context
  const node = ast.codegenNode.content
  push(`return '${node}'`)
}
