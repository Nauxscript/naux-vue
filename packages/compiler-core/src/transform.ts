export function transform(root, options: Record<any, any> = {}) {
  const { nodeTransforms } = options
  if (nodeTransforms) {
    for (let i = 0; i < nodeTransforms.length; i++)
      nodeTransforms[i](root)
  }
  traverseNode(root, options)
}

function traverseNode(node, options) {
  if (!node.children)
    return
  for (let i = 0; i < node.children.length; i++)
    transform(node.children[i], options)
}
