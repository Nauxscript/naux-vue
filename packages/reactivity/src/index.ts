export {
  reactive,
  readonly,
  shallowReadonly,
  isReactive,
  isReadonly,
  isProxy,
} from './reactive'
export { effect, ReactiveEffect } from './effect'
export { ref, proxyRefs } from './ref'
export { computed } from './computed'
