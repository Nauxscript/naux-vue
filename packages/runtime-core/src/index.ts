export { createApp } from './createApp'
export { createVNode } from './vnode'
export { h } from './h'
