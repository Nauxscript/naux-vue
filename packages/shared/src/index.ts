export const extend = Object.assign

export const isObject = (target: any) => target !== null && typeof target === 'object'

export const hasChanged = (value: any, newValue: any) => !Object.is(value, newValue)
