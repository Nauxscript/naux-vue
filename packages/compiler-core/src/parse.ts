import { NodeTypes } from './ast'

type Context = ReturnType<typeof createParseContext>

export const baseParse = (content: string) => {
  const context = createParseContext(content)
  return createRoot(parseChildren(context))
}

function parseChildren(context: Context) {
  const nodes: any[] = []
  let node

  const { source } = context

  if (source.startsWith('{{')) {
    // interpolation
    node = parseInterpolation (context)
  }
  else if (source.startsWith('<')) {
    // element
  }
  else {
    // text
    node = parseText(context)
  }
  nodes.push(node)
  return nodes
}

function createRoot(children) {
  return { children }
}

function createParseContext(content: string) {
  return { source: content }
}

function parseText(context: Context) {
  return {
    type: NodeTypes.TEXT,
    content: context.source,
  }
}

function parseInterpolation(context: Context): any {
  const openDelimiter = '{{'
  const closeDelimiter = '}}'

  // finde open delimiter index
  const closeDelimiterIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)

  // get content
  const content = context.source.slice(openDelimiter.length, closeDelimiterIndex).trim()

  // remove the interpolation
  context.source = context.source.slice(closeDelimiterIndex + closeDelimiter.length)

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  }
}
