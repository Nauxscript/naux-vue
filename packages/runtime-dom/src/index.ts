import { createRenderer } from '@naux-vue/runtime-core'

function createElement(type) {
  return document.createElement(type)
}

function patchProp(el, key, oldVal, newVal) {
  if (key.match(/^on[A-Z]/)) {
    const eventName = key.slice(2).toLowerCase()
    el.addEventListener(eventName, newVal)
  }
  else {
    // eslint-disable-next-line no-console
    console.log(oldVal)
    // eslint-disable-next-line no-console
    console.log(newVal)
    // eslint-disable-next-line no-console
    console.log('------------')
    if (oldVal !== newVal) {
      if (newVal === null || newVal === undefined)
        el.removeAttribute(key, newVal)
      else
        el.setAttribute(key, newVal)
    }
  }
}

function insert(el, parent) {
  parent.append(el)
}

function createTextNode(text) {
  return document.createTextNode(text)
}

function setElementText(el, text) {
  // what's the difference between textContent and innerText (and innerHTML)
  // the most important defference is, use innerText to change a dom content will trigger reflow, but textContent woundn't.
  // refer to https://juejin.cn/post/7044436311033249805
  el.textContent = text
}

function remove(el) {
  const parent = el.parentNode
  parent && parent.removeChild(el)
}

const renderer = createRenderer({
  createElement,
  patchProp,
  insert,
  createTextNode,
  setElementText,
  remove,
}) as any

export function createApp(...args) {
  return renderer.createApp(...args)
}

export * from '@naux-vue/runtime-core'
