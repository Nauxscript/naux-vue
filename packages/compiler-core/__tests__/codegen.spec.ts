import { describe, it } from 'vitest'
import { transform } from '../src/transform'
import { generate } from '../src/codegen'
import { baseParse } from './../src/parse'
describe('codegen', () => {
  it('basic string', () => {
    const ast = baseParse('hello, nau')
    transform(ast)
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })
})
