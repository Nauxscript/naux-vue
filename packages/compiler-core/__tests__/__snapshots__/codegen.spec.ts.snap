// Vitest Snapshot v1

exports[`codegen > basic string 1`] = `"return function render(_ctx,_cache){return 'hello, naux'}"`;
